### ISSUE TITLE
High warmer time and threads reported when no warmers exist

### ISSUE THREAD

**Issue by allthedrones** [elastic/elasticsearch#14343]
This is happening on an elasticsearch cluster with a relatively large number of indices (~2500), where **zero** indices have warmers defined. I most recently noticed that on a node that had only been running for 4 hours, its node stats reported 3.8h of processing time spent on warmers. Is there some other operation other than true index warmers that are being reported in this category? On this same node, the thread pool stats reported `completed` values of about 1000:1 for warmer threads vs other operations (search, generic, etc).

Relevant sections from the *indices* and *thread_pool* sections of `_nodes/stats`:

```json
"warmer": {
      "current": 0,
      "total": 34628,
      "total_time": "3.8h",
      "total_time_in_millis": 13805427
    }
```
[full indices section of _nodes/stats](https://gist.github.com/allthedrones/f27b1e1471c031fe7c52#file-node-stats-indexes-json)

```json
"warmer": {
      "threads": 4,
      "queue": 0,
      "active": 0,
      "rejected": 0,
      "largest": 4,
      "completed": 24963774
    }
```
[full thread_pool section of _nodes/stats](https://gist.github.com/allthedrones/f27b1e1471c031fe7c52#file-node-stats-threadpool-json)

**Comment by jasontedor** [comment]
This is likely from the top warmer. If you want to verify, you enable trace logging for `index.warmer` and look for `top warming` in the logs.

Have you verified that this is harming your performance in any meaningful way? If you have, you can disable this by setting `index.warmer.enabled` to `false`.You can set this cluster wide, or on an individual index (it sounds like you want this cluster wide). You should be able to verify that this had the desired effect by requesting

```
_cat/thread_pool\?v\&h=warmer.type,\
warmer.active,\
warmer.size,\
warmer.queue,\
warmer.queueSize,\
warmer.rejected,\
warmer.largest,\
warmer.completed,\
warmer.min,\
warmer.max,\
warmer.keepAlive
```

before changing the setting, then changing the setting, indexing some documents, and then checking the output of the above request again. After `warmer.keepAlive` expires, all of the threads in this thread pool should be terminated.

**Comment by allthedrones** [comment]
Thanks for the clarification. We've noticed some *correlation* between high warmer activity and very high (~90%) heap, with virtually all of the heap usage promoted up to the "old" pool. Obviously this doesn't necessarily imply a causal link. Is there any documentation on what the top warmer does? That might lead us to some *other* behavior or activity which might be a more likely culprit in the heap growth. I couldn't find anything by searching for `elasticsearch "top warmer"`. Thanks again for the quick response.

**Comment by allthedrones** [comment]
We were able to verify that disabling warmers (even with no warmers defined) caused heap usage to drop from ~90% to ~30%, with no discernable difference in performance ... so there is definitely a bug or undocumented performance consideration here. This was cluster wide for 8 data nodes with 28G of heap space, so it's an enormous impact.

**Comment by clintongormley** [comment]
The top warmer is used to warm up data structures that are per shard, not per segment.  eg parent-child global ordinals, any fields with global_ordinals set to eager, etc.

If you're not seeing any performance impact, then perhaps you have mappings like the above which aren't actually being used?

**Comment by ispringer** [comment]
We are running ES 1.7.2 and saw the exact same issue as @allthedrones - unexplained abnormally high CPU and heap usage. All of the hot threads were warmer threads. We turned on trace logging for `index.warmer` and saw that the top warmer was running 2-3 times per second. When we set `index.warmer.enabled` to false globally, the CPU and heap usage quickly dropped to normal levels, and there was no discernible impact on query or indexing performance. In my opinion, this is a very serious bug, and I think this issue should be reopened.
